Assignment5.java

```diff
@@ -0,0 +1,198 @@
+     */
+    private static void addAllCharacters(String s, Set<Character> set) {
+        for (int i = 0; i < s.length(); i++) {
+            set.add(s.charAt(i));
+        }
+    }
+
+    private static class NoDifferenceException extends Exception {
+
+    }
+
+    /**
+     * Finds the index at which s1 and s2 start to differ.
+     *
+     * @param s1 string number one.
+     * @param s2 string number two.
+     * @return index s1 and s2 start to differ.
+     * @throws NoDifferenceException if s1 and s2 are not different in an index of the shorter one.
+     */
+    private static int findFirstDifferingIndex(String s1, String s2) throws NoDifferenceException {
+        int bound = Math.min(s1.length(), s2.length());
+        int index = 0;
+
+        while (index < bound && s1.charAt(index) == s2.charAt(index)) {
+            index++;
+        }
+
+        if (index == bound) {
+            throw new NoDifferenceException();
+        }
+
+        return index;
+    }
+
+    /**
+     * Creates vertices with values from and to, and an edge between those in graph.
+     *
+     * @param from  value of the vertex the edge will be coming from.
+     * @param to    value of the vertex the edge will be coming to.
+     * @param graph graph containing the vertices.
+     */
+    private static void createEdgeBetweenVertices(char from, char to, Map<Character, Vertex> graph) {
+        if (!graph.containsKey(from)) {
+            graph.put(from, new Vertex(from));
+        }
+
+        if (!graph.containsKey(to)) {
+            graph.put(to, new Vertex(to));
+        }
+
+        graph.get(from).addEdge(graph.get(to));
+
+        graph.get(to).incrementindegree();
+    }
+
+    /**
+     * Performs topological sort on graph.
+     *
+     * @param graph graph containing the vertices.
+     * @return topologically sorted set of characters.
+     */
+    private static Set<Character> topologicalSort(Map<Character, Vertex> graph) {
+        Set<Character> set = new LinkedHashSet<>();
+
+        if (graph.isEmpty()) {
+            return set;
+        }
+
+        Vertex[] vertices = new Vertex[graph.size()];
+
+        int index = 0;
+
+        for (Map.Entry<Character, Vertex> entry : graph.entrySet()) {
+            vertices[index] = entry.getValue();
+            index++;
+        }
+
+        Queue<Vertex> queue = new LinkedList<>();
+
+        while (true) {
+            for (Vertex v : vertices) {
+                if (v.getindegree() == 0) {
+                    queue.add(v);
+                    v.decrementindegree();
+                }
+            }
+
+            if (queue.isEmpty()) {
+                break;
+            }
+
+            while (!queue.isEmpty()) {
+                Vertex v = queue.poll();
+                set.add(v.getValue());
+
+                for (Vertex el : v.getEdges()) {
+                    el.decrementindegree();
+                }
+            }
+        }
+
+        return set;
+    }
+}
+
+class Vertex {
```

Comment:
Or, to be more precise: this particular one should either be in a separate file or should be inner class of `Assignment5`.